test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -51,258 +53,458 @@ var (
-	if err != nil {
-		b.Fatal(err)
-	}
-	if err := nodePreparer.PrepareNodes(); err != nil {
-		b.Fatal(err)
-	}
-	defer nodePreparer.CleanupNodes()
-
-	total := 0
-	for _, p := range test.InitPods {
-		if err := createPods(setupNamespace, p, clientset); err != nil {
-			b.Fatal(err)
-		}
-		total += p.Num
-	}
-	if !test.SkipWaitUntilInitPodsScheduled {
-		if err := waitNumPodsScheduled(b, total, podInformer, setupNamespace); err != nil {
-			b.Fatal(err)
-		}
-	}
+	numPodsScheduledPerNamespace := make(map[string]int)
+	numNodes := 0
 
-	// start benchmark
 	b.ResetTimer()
+	for opIndex, op := range w.Ops {
+		switch realOp := op.realOp.(type) {
+		case *createNodesOp:
+			nodePreparer, err := getNodePreparer(fmt.Sprintf("node-%d-", opIndex), realOp, clientset)
+			if err != nil {
+				b.Fatal(err)
+			}
+			if err := nodePreparer.PrepareNodes(numNodes); err != nil {
+				b.Fatal(err)
+			}
+			if numNodes == 0 {
+				// Schedule a cleanup at most once. The CleanupNodes function will list
+				// and delete *all* nodes.
+				// TODO(#93794): make CleanupNodes only clean up its own nodes to make
+				// this more intuitive?
+				defer nodePreparer.CleanupNodes()
+			}
+			numNodes += realOp.CreateNodes
+
+		case *createPodsOp:
+			var namespace string
+			if realOp.Namespace != nil {
+				namespace = *realOp.Namespace
+			} else {
+				namespace = fmt.Sprintf("namespace-%d", opIndex)
+			}
+			var stopCh chan struct{}
+			var collectors []testDataCollector
+			if realOp.CollectMetrics {
+				b.StartTimer()
+				stopCh = make(chan struct{})
+				collectors = getTestDataCollectors(podInformer, fmt.Sprintf("%s/%s", b.Name(), namespace), namespace, tc.MetricsCollectorConfig)
+				for _, collector := range collectors {
+					go collector.run(stopCh)
+				}
+			}
+			if err := createPods(namespace, realOp, clientset); err != nil {
+				b.Fatal(err)
+			}
+			finish := func() {
+				if err := barrierOne(podInformer, b.Name(), namespace, realOp.CreatePods); err != nil {
+					b.Fatal(err)
+				}
+				if realOp.CollectMetrics {
+					close(stopCh)
+					mu.Lock()
+					for _, collector := range collectors {
+						dataItems = append(dataItems, collector.collect()...)
+					}
+					mu.Unlock()
+				}
+			}
+			if realOp.SkipWaitToCompletion {
+				// Only record those namespaces that may potentially require barriers
+				// in the future.
+				if _, ok := numPodsScheduledPerNamespace[namespace]; ok {
+					numPodsScheduledPerNamespace[namespace] += realOp.CreatePods
+				} else {
+					numPodsScheduledPerNamespace[namespace] = realOp.CreatePods
+				}
+				go finish()
+				// We can't stop the timer after the pods actually get scheduled
+				// because a different timer may have been started and stopped by then.
+				// AFAIK, there is no way to add the time elapsed to the benchmark's
+				// timer concurrently without using StartTimer and StopTimer.
+				b.StopTimer()
+			} else {
+				finish()
+				b.StopTimer()
```

Comment:
are those two StopTimer calls still necessary? I think not.

Reply:
Oops, sorry about that, must've missed those two. Thanks for catching! Done.